lib/mutant/ast/regexp/transformer.rb

```diff
@@ -0,0 +1,76 @@
+module Mutant
+  module AST
+    module Regexp
+      class Transformer
+        include AbstractType
+
+        REGISTRY = TypeRegistry.new
+
+        def self.lookup(type)
+          REGISTRY.lookup(type)
+        end
+
+        def self.register(type)
+          REGISTRY.add(type, self)
+        end
+        private_class_method :register
+
+        def self.to_ast(expression)
+          self::ExpressionToAST.call(expression)
+        end
+
+        def self.to_expression(node)
+          self::ASTToExpression.call(node)
+        end
+
+        class ExpressionToAST
+          PREFIX = :regexp
+
+          include Concord.new(:expression), Procto.call, AST::Sexp, AbstractType, Adamantium
+
+        private
+
+          def children
+            expression.expressions.map(&Regexp.method(:to_ast))
+          end
+
+          def quantify(node)
+            return node unless expression.quantified?
+
+            Quantifier.to_ast(expression.quantifier).append(node)
+          end
+
+          def ast(*children)
+            s(ast_type, *children)
+          end
+
+          def ast_type
+            :"#{PREFIX}_#{expression.token}_#{expression.type}"
+          end
+        end
+
+        class ASTToExpression
+          include Concord.new(:node), Procto.call, AbstractType, Adamantium
+
+        private
+
+          def subexpressions
+            node.children.map(&Regexp.public_method(:to_expression))
+          end
+        end
+
+        module LookupTable
+        private
+
+          def expression_token
+            self.class::TABLE.fetch(node.type).first
```

Comment:
Instead of `#first` here, and `#last` below, which are rather weak assertions about the shape of the code, what about doing something like `Quantifier` below and creating a simple class to represent the expression. You can call `#token` and `#regexp_class` on the object or something.